### ISSUE TITLE
PCA calculations not using entire MMAP space

### ISSUE THREAD

**Issue by lukaszbachman** [deeplearning4j/deeplearning4j#8865]
I'm trying to run a simple experiment with ND4J where PCA calculations will be offloaded to an MMAP. In production I plan to use this on large matrices (ex: 30_000 columns, 20_000 rows). For testing I'm using something much smaller to verify if I'm using the API in a proper way. 
I have read relevant sections related to `INDArrays` and memory usage on DeepLearning4j.org and nd4j.org, but I'm still not getting something.

I have attempted several tests before wrt. the MMAP usage and I noticed that I don't really need to set large `-Xmx` and `-Dorg.bytedeco.javacpp.maxbytes` values. I am using respectively 50mb and 70mb and that allowed me to successfully create MMAPs as large as 8GBs filled with `Nd4j.rand(...)` from the very first byte up to the very end of the file. 
I also noticed that if I attempt to create a bigger array that won't fit into the MMAP, I may end up with JVM Crash (see #8864) or the MMAP won't be used at all (thanks to @raver119 for giving me a hint about spilled arrays). 

However what I'm looking at right now looks a bit odd. First, an example to reproduce:
```
public static void main(String[] args) {
    WorkspaceConfiguration mmap = WorkspaceConfiguration.builder()
      .initialSize(100 * 1024L * 1024L) // 100mbs
      .tempFilePath("/tmp/my.mmap")
      .policyLocation(LocationPolicy.MMAP)
      .build();
    try (MemoryWorkspace ws = Nd4j.getWorkspaceManager().getAndActivateWorkspace(mmap, "M2")) {
      int samplesCount = 1000;
      int featuresCount = 500;

      INDArray matrix = Nd4j.rand(DataType.FLOAT, samplesCount, featuresCount);
      INDArray dup = matrix.dup();
      INDArray factor = PCA.pca_factor(matrix, 0.95, false); // EXCEPTION THROWN HERE
      INDArray reducedMatrix = dup.mmul(factor);

      double[][] reduced = reducedMatrix.toDoubleMatrix();
      System.out.println("Reduced = " + reduced.length + " x " + reduced[0].length);
    }
  }
```

Running this code I ended up with this exception:
```
Exception in thread "main" java.lang.OutOfMemoryError: Cannot allocate new LongPointer(1): totalBytes = 264, physicalBytes = 125M
...
Caused by: java.lang.OutOfMemoryError: Physical memory usage is too high: physicalBytes (125M) > maxPhysicalBytes (120M)
...
```

I jumped to hex editor and checked where is the last non-zero byte stored, and it was close to offset `004c5ae0`. This means that around 1_251_000 floats was stored. But it also means that the file had also room for 24_963_403 more (if I'm doing my math right: `(0x6400000−0x004C5AD4) / 4`).
So it looks to me that the MMAP had plenty of room left (96%) but for some reason I ran into OOME.

I'm running this with :
```
-Xmx50m -Dorg.bytedeco.javacpp.maxbytes=70m
```
I know that likely changing memory settings could bail me out here, but I need to understand what the problem is in order to avoid similar issues in commercial deployments.

#### Version Information
* ND4J 1.0.0-beta6, CPU backend
* Ubuntu 19.04

#### Additional Information

https://gist.github.com/lukaszbachman/57bb2b17b6f4e8b7e00ce5b8935c30a2

![mmap-end](https://user-images.githubusercontent.com/676753/79981395-e635a600-84a4-11ea-999c-b883c152ec08.png)

**Comment by saudet** [comment]
The OOM you're getting is from JavaCPP checking the total physical memory the process is using and trying to make sure it stays under that limit by calling `System.gc()` and `malloc_trim()`. If this is something you don't need to monitor, that is, you're already making sure in other ways that your process is never going to exceed some given limits, this check can be disabled by setting maxPhysicalBytes to 0:
http://bytedeco.org/javacpp/apidocs/org/bytedeco/javacpp/Pointer.html#maxPhysicalBytes

**Comment by lukaszbachman** [comment]
That is good to know indeed! But I still don't get how come JavaCPP is deciding that my physical usage is too big in that case. After all I'm using MMAP, so everything I do inside this workspace should be directly mapped to that file (and not being accounted for as `physicalBytes`) and it had plenty of room left. I think I'm still missing something here.

**Comment by saudet** [comment]
Apparently Linux counts it in the "resident set size". I'm sure it gets cached in memory there anyway.

**Comment by lukaszbachman** [comment]
But then how to configure the memory settings so that the application won't trip on large matrices sent to PCA? Say I have a machine with 32gbs of ram and my regular usage looks like this:

- 2 GB - reserved for OS
- 10 GB - Xmx reserved for application during normal usage

Assuming that PCA could require as much as another 30 GBs of memory, how should I configure the parameters to avoid running out of memory?

**Comment by saudet** [comment]
I assume Linux can potentially cache the whole file to memory, so I guess "physicalBytes" could go up to 30 GB + Xmx + maxBytes...

**Comment by treo** [comment]
In principle loading the whole file into memory is what you want, the benefit of mmaping is that it can spill to disk for things that aren't required quite as often. This makes it ideal for working with frozen word vectors, where some words will be needed more often than others, and the OS needs to only keep a few hot pages actually in memory. 

I'm a bit rusty on the specific of PCA calculations and their memory access patterns, but I think that the way you are using it mmap as swap, you might as well just set up your system with more swap in the first place (unless that isn't possible for you for some reason).

**Comment by lukaszbachman** [comment]
Hhhmm but is it possible to request more `maxbytes` than actual RAM installed? Because in my case the "extra stuff" will be peanuts comparing to the features matrix that goes into the PCA algorithm.

Hhmm I just tried doing this on my 16GB laptop and it didn't complain...
```
 System.setProperty("org.bytedeco.javacpp.maxbytes", "20G");
 System.setProperty("org.bytedeco.javacpp.maxphysicalbytes", "40G");
```

**Comment by saudet** [comment]
Yes, of course, that works, and for "unlimited", simply use 0.

**Comment by lukaszbachman** [comment]
One more question related to this spilling of arrays - say that my MMAP has enough free space to accomodate 1GB of data, but I try to create a matrix which would use 3GBs. What happens during the moment of allocation? Since 3GB > 1GB then entire 3GB spills out of the workspace? Or perhaps 1GB is allocated into a file and the remaining 2GBs spill to RAM?

**Comment by raver119** [comment]
No, it's impossible to split single INDArray over multiple memory chunks, so it's all-or-nothing. Array either fits into workspace, or spills.

**Comment by lukaszbachman** [comment]
Ok, that sounds logical. Still, it looks to me that the MMAP had enough room to accommodate all calculations, yet it seems that it wasn't used. Am I making any sense here or maybe I'm missing something obvious?

**Comment by lukaszbachman** [comment]
So I ran more tests and my initial code ran fine if I bumped the MMAP size to 16GBs. It didn't throw any OOME. I tested the produced binary file and noted the hex address of the last non-null byte:

`0092c250: 7b6e ce3e 70ea 133f 00ff 603b 0000 0000  {n.>p..?..`;....`

So that's `0x92c250 == 9618000`. If I divide by 4 it means that the last written float was at position:
`9618000 / 4 + 5 = 2_404_505`. So running PCA on `M=1000` and `N=500` needed 2.4mln of floats (or <10MBs) to be allocated in memory. 

It really looks to me that something is leaking outside of my MMAP memory workspace.

When MMAP is smaller then this code often fails on this line from `PCA.pca_factor()`

```
INDArray factor = Nd4j.createUninitialized(A.dataType(), new long[]{n, k}, 'f');
        for (int i = 0; i < k; i++) {
            factor.putColumn(i, V.getColumn(i)); <-- HERE
        }
```

Maybe there is some leak there?

**Comment by saudet** [comment]
@lukaszbachman If you've disabled checks by JavaCPP, native memory may be "leaking" if the garbage collector doesn't get run. Try to call `System.gc()` once in a while to see if that works.

**Comment by agibsonccc** [comment]
Closing — this issue is over 5 years old with no recent activity. The codebase has been significantly rewritten. If this is still relevant, please open a new issue.
